### ISSUE TITLE
Fixture TestName for fixtures: to template or not to template?

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2675]
Before we ship `TestFixtureData.SetName` or `TestFixtureData.SetArgDisplayNames`:

- Should the user-specified **test fixture name** be interpreted as a literal name, or should it follow the name templating spec the way test case names do?

It's so easy to lose track of things. I've been scratching my head over https://github.com/nunit/nunit/pull/2538 being unreviewed for six weeks. 🕸 Charlie [gave me a clue](https://github.com/nunit/nunit/pull/2538#issuecomment-359631175) that makes sense to me. As I do the work to separate my PR, would you please chime in and help decide which way we are going?

I'm clearly biased towards making TestFixtureData.TestName and TestCaseData.TestName both template. The arguments I would put forth have to do with simplicity for end users and elegance on the inside of NUnit, blah blah blah. I don't want to “uniquify” TestFixtureData.TestName from TestCaseData.TestName. I'll try not to bore you further on the basis of personal preference. ☝

#### Less wishy-washy

`TestFixtureData.TestName`, from a users' perspective, is all about controlling how the fixture parameters display. They display in the name of the fixture and the full names of the tests underneath it. You're only likely to reach for it if you end up with duplicate test names or cryptic or ugly test names– in every case due to the autogenerated parameter display text.
`TestCaseData.TestName` likewise– you're only likely to use it as a means of overriding the default test method parameter display. You don't have child tests to worry about, but in all other ways the user thinks about customizing the test _method_ parameter display and customizing the fixture _constructor_ parameter display in the same way and for the same reasons.

This is what I care about. The parallelism in the use of constructor parameters and the use of method parameters should be reflected in the API.

#### Further observations

- If we do decide to make a difference in how we treat TestName, depending when it's inherited by `TestFixtureData` or by `TestCaseData`, we'll have to code name-related things twice.
  A concrete example is `SetArgDisplayNames`. (In real projects, I'm already waiting for it for constructor parameters in exactly the same way that I need it for method parameters.) If `TestFixtureData.TestName` does **not** template, `SetArgDisplayNames` will have to be responsible to use TestNameGenerator directly and overwrite the literal `TestName` immediately. This is unlike `TestCaseData` where the argument display names are saved for later so that `{a:40}` (or any runner's pattern override) does its job much lower. We can pull a couple of things inside out to make the internals more similar and smooth over the external difference, but maybe this consideration should inform our API decision.

- We have a `{c}` replacement token in the template spec, but no one uses that in `TestCaseData.TestName` because it's already in the full name.

Disagreeing opinions are welcome! :+1:

**Comment by CharliePoole** [comment]
Runners like the adapter are already dealing with test case names, which may be changed by the user - not always with a lot of success, however. Up to now, they have been able to assume that the full name of the containing fixture is __not__ subject to modification. If we change that, I'm not sure how we would be able to interact with Visual Studio.

It would be possible to use a template approach to generate the fixture name but __not__ allow the user to change it. However, we would then be doing the same thing we do now in a more complicated way for the sake of some sort of internal elegance.

**Comment by CharliePoole** [comment]
Yet another reason not to mix things in a PR. You are now telling us that we have to decide something that's very hard and tricky to decide in order to fix something that's relatively easy. I realize that you have already invested a lot of time in this, but trying to make all the decisions up front so that you don't have to rewrite code is usually a losing game, at least in my experience.

**Comment by CharliePoole** [comment]
My preferred approach here is to remove TestName from TestFixtureData. I put it in experimentally and it caused problems. There are two ways to go: do some complicated stuff to make it work or remove it. As someone who at first wanted it, I still don't want it that badly. It can be removed entirely without issue because it has never been released and nobody has ever asked for it.

One thing I like to keep in mind is that we can always add features. What we can't seem to do very easily is take them out once people are using them. Let's not add speculative features.

**Comment by jnm2** [comment]
I don't think it should be possible to override full names, only name segments. That way runners work (as well as my intuition).

Yes, I'm learning on my feet. Your advice is helpful!

I hope this is not too tricky of a question, though, because I already need SetArgDisplayNames on test fixtures and will be sorely disappointed if the ability to control the fixture name segment is vetoed. My options are either to listen to the team and work to implement *this* issue, and then add TestFixtureData.SetArgDisplayNames on top of that, or to stop using constructor parameters with fixtures in general. This is already real-world for me; no speculation here. If I'm using parameters, customization is a must, or so says my real-world difficulties.

**Comment by CharliePoole** [comment]
It seems as if it should be possible to set the arg names without using templating. Isn't it?

**Comment by jnm2** [comment]
Ah, I understood you to be saying that you prefer no name customization for fixtures. You're okay with having SetArgDisplayNames without SetName or TestName?

**Comment by CharliePoole** [comment]
Absolutely. It's a feature that I added, which didn't work out. Thought I had said that and that we removed it. Perhaps not.

**Comment by jnm2** [comment]
Okay. The tricky thing is that TestName is still public and settable on TestFixtureData, inherited from TestParameters. We might consider moving it to TestCaseParameters so that we don't have to think about wheat to do when people set it.

**Comment by jnm2** [comment]
Okay. The tricky thing is that TestName is still public and settable on TestFixtureData, inherited from TestParameters. We might consider moving it to TestCaseParameters so that we don't have to think about what to do when people set it.

Of course, that's a breaking change if anyone is using `TestName` by dotting off `TestParameters` or any fixture-specific inheriting class or doing `new TestFixtureData(...) { TestName = "" }` which is documented public API. But it can't have been working in a helpful way so that's not something we'll have to worry about, right?

**Comment by CharliePoole** [comment]
Bummer! Our intention back in November, was to keep this all internal till we made further decisions. PR #2554 made `TestFixtureData.SetName` internal but we missed `TestParameters.TestName`. My tendency would be to go ahead and break this since it's really not implemented fully.

Alternatively, we could keep it without allowing use of a template. It's not very useful, but if somebody found it so there seems to be no harm.

**Comment by jnm2** [comment]
#2554 made sure there was no change between release 3.8.1 and 3.9. `TestParameters.TestName` has been public and settable since NUnit 2.9.7 (https://github.com/nunit/nunit/commit/5e799831dc2ff1682cdd6228d9807592dd160c15). But I'm still pretty confident no one uses it.

> Alternatively, we could keep it without allowing use of a template.

That's one side of the question at the top. My preference being to make it behave consistently no matter whether it's inherited into a TestCaseData or TestFixtureData.

**Comment by jnm2** [comment]
~~But I'm still pretty confident no one uses it.~~

What am I saying; I use it in production tests that have parameterized fixtures...
